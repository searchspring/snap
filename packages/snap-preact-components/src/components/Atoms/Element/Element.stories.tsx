import { h } from 'preact';

import { ArgsTable, PRIMARY_STORY } from '@storybook/addon-docs/blocks';

import { Element, ElementProps } from './Element';
import { componentArgs } from '../../../utilities';
import Readme from '../Dropdown/readme.md';

export default {
	title: `Atoms/Element`,
	component: Element,
	parameters: {
		docs: {
			page: () => (
				<div>
					<Readme />
					<ArgsTable story={PRIMARY_STORY} />
				</div>
			),
		},
	},
	argTypes: {
		type: {
			description: 'Determines the type of element to render. ',
			type: { required: true },
			table: {
				type: {
					summary: 'string',
				},
			},
			control: {
				type: 'select',
				options: ['div', 'span', 'p', 'label'],
			},
		},
		content: {
			description: 'Content to be displayed in Element',
			table: {
				type: {
					summary: 'string',
				},
			},
			control: { type: 'text' },
		},
		attributes: {
			description: 'Attributes to be added to the element. (onClick not allowed)',
			defaultValue: {},
			table: {
				type: {
					summary: 'object',
				},
			},
			control: {
				type: 'object',
			},
		},
		...componentArgs,
	},
};

export const Div = (args: ElementProps) => <Element {...args} />;
Div.args = {
	type: 'div',
	content: "I'm a div",
};

export const Span = (args: ElementProps) => <Element {...args} />;
Span.args = {
	type: 'span',
	content: "I'm a span",
};

export const Label = (args: ElementProps) => <Element {...args} />;
Label.args = {
	type: 'label',
	content: "I'm a label with attributes",
	attributes: {
		attribute1: 'sku',
		attribute2: 'link',
	},
};
