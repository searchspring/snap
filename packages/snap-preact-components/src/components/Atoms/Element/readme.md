# Element

Renders an HTML element of your choice. This component can be used for adding basic strings to your layouts, or for adding specific elements for third parties to target and inject into. 

## Usage

### type
The required `type` prop specifies the element type. This can be any of the supported element types - `div`, `span`, `p` or `label`. 

```jsx
<Element type={"div"} />
```
 
### content
The `content` prop specifies the content to be rendered within the element. This only accepts a string.

```jsx
<Element type={"span"} content={"Hello World!"} />
```

### attributes
The `attributes` prop specifies additional attributes to be spread onto the element. This can be any attribute/value you may need, however `onClick` attributes are <b>not</b> supported. 

```jsx

const attributes = {
    sku: '1234',
    id: 'someRatingWidgetID',
    ['data-name']: "product1234",
}

<Element type={'div'} content={"3rd party ratings append here!"} attributes={attributes}/>
```
