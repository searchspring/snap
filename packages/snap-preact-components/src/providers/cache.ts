export { CacheProvider } from '@emotion/react';
import createCache from '@emotion/cache';

export const cache = createCache({
	key: 'ss',
	prepend: true,
});
