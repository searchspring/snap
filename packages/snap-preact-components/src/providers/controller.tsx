import { jsx } from '@emotion/react';
import preact, { h, createContext, ComponentChildren, ComponentType } from 'preact';
import { useContext } from 'preact/hooks';
import type { AbstractController } from '@searchspring/snap-controller';

const ControllerContext = createContext<null | AbstractController>(null);

export const ControllerProvider = ({ children, controller }: { children: ComponentChildren; controller: AbstractController }) => {
	return <ControllerContext.Provider value={controller}>{children}</ControllerContext.Provider>;
};

export const useController = () => useContext(ControllerContext);

export const withController = (Component: ComponentType) => (props: any) => {
	return <Component controller={useController()} {...props} />;
};
