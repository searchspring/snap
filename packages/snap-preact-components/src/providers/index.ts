export * from './store';
export * from './theme';
