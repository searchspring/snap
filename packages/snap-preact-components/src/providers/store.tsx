import preact, { h, createContext, ComponentChildren, ComponentType } from 'preact';
import { useContext } from 'preact/hooks';
import type { AbstractStore } from '@searchspring/snap-store-mobx';

const StoreContext = createContext<null | AbstractStore>(null);

export const StoreProvider = ({ children, store }: { children: ComponentChildren; store: AbstractStore }) => {
	return <StoreContext.Provider value={store}>{children}</StoreContext.Provider>;
};

export const useStore = () => useContext(StoreContext);

export function withStore<C extends ComponentType>(Component: C): C {
	return ((props: any) => <StoreContext.Consumer>{(store) => <Component {...props} store={store} />}</StoreContext.Consumer>) as C;
}
