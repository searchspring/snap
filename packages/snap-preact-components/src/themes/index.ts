export * from './pike';
export * from './bocachica';
