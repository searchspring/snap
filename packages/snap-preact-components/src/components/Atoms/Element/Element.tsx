/** @jsx jsx */
import { Fragment, h } from 'preact';
import { jsx, css } from '@emotion/react';
import classnames from 'classnames';
import { observer } from 'mobx-react-lite';
import { Theme, useTheme, CacheProvider } from '../../../providers';
import { ComponentProps, StylingCSS } from '../../../types';
import { mergeProps } from '../../../utilities';

const CSS = {
	Element: ({}: Partial<ElementProps>) => css({}),
};

export const Element = observer((properties: ElementProps): JSX.Element => {
	const globalTheme: Theme = useTheme();
	const defaultProps: Partial<ElementProps> = {};

	const props = mergeProps('element', globalTheme, defaultProps, properties);

	const { type, content, attributes, disableStyles, className, style, styleScript } = props;

	const styling: { css?: StylingCSS } = {};
	const stylingProps = props;

	if (styleScript && !disableStyles) {
		styling.css = [styleScript(stylingProps), style];
	} else if (!disableStyles) {
		styling.css = [CSS.Element(stylingProps), style];
	} else if (style) {
		styling.css = [style];
	}

	const args = {
		...styling,
		...attributes,
		className: classnames('ss__element', className),
		children: content,
		onClick: undefined,
	};

	return type ? (
		<CacheProvider>
			{type == 'div' && <div {...args}></div>}

			{type == 'span' && <span {...args}></span>}

			{type == 'p' && <p {...args}></p>}

			{type == 'label' && <label {...args}></label>}
		</CacheProvider>
	) : (
		<Fragment />
	);
});

export interface ElementProps extends ComponentProps {
	type: 'div' | 'span' | 'p' | 'label';
	content?: string;
	attributes?: {
		[name: string]: string | number | undefined;
		//no onclicks allowed
		onClick?: undefined;
	};
}
