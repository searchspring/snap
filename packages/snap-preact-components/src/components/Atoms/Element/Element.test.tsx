import { h } from 'preact';

import { render } from '@testing-library/preact';

import { Element } from './Element';
import { ThemeProvider } from '../../../providers';
import themes from '../../../themes';

describe('Element Component', () => {
	const text = 'some text';

	Object.keys(themes || {}).forEach((themeName) => {
		it(`uses ${themeName} theme`, () => {
			const theme = themes[themeName as keyof typeof themes];
			const rendered = render(<Element theme={theme} type={'div'} content={text} />);
			expect(rendered.asFragment()).toMatchSnapshot();
		});
	});
	it('renders as div', () => {
		const rendered = render(<Element type={'div'} content={text} />);

		const element = rendered.container.querySelector('div.ss__element');
		expect(element).toBeInTheDocument();
		expect(element).toHaveTextContent(text);
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('renders as span', () => {
		const rendered = render(<Element type="span" content={text} />);

		const element = rendered.container.querySelector('span.ss__element');
		expect(element).toBeInTheDocument();
		expect(element).toHaveTextContent(text);
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('renders as label', () => {
		const rendered = render(<Element type="label" content={text} />);

		const element = rendered.container.querySelector('label.ss__element');
		expect(element).toBeInTheDocument();
		expect(element).toHaveTextContent(text);
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('renders as p', () => {
		const rendered = render(<Element type="p" content={text} />);

		const element = rendered.container.querySelector('p.ss__element');
		expect(element).toBeInTheDocument();
		expect(element).toHaveTextContent(text);
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('doesnt render without type prop', () => {
		// @ts-ignore
		const rendered = render(<Element />);

		const element = rendered.container.querySelector('.ss__element');
		expect(element).not.toBeInTheDocument();
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('renders with attributes', () => {
		const attributes = {
			attribute1: 'one',
			attribute2: 'two',
			attribute3: 3,
		};
		const rendered = render(<Element type="div" content={text} attributes={attributes} />);

		const element = rendered.container.querySelector('div.ss__element');
		expect(element).toBeInTheDocument();
		expect(element).toHaveTextContent(text);

		expect(element).toHaveAttribute('attribute1', attributes.attribute1);
		expect(element).toHaveAttribute('attribute2', attributes.attribute2);
		expect(element).toHaveAttribute('attribute3', attributes.attribute3.toString());
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('ignores onClick attribute', () => {
		const attributes = {
			onClick: () => {},
			attribute2: 'two',
			attribute3: 3,
		};
		// @ts-ignore
		const rendered = render(<Element type="div" content={text} attributes={attributes} />);

		const element = rendered.container.querySelector('div.ss__element');
		expect(element).toBeInTheDocument();
		expect(element).toHaveTextContent(text);

		expect(element).not.toHaveAttribute('onClick');
		expect(element).toHaveAttribute('attribute2', attributes.attribute2);
		expect(element).toHaveAttribute('attribute3', attributes.attribute3.toString());
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('renders with className prop', () => {
		const className = 'classy';
		const rendered = render(<Element type={'div'} className={className} />);

		const element = rendered.container.querySelector(`.ss__element.${className}`);
		expect(element).toBeInTheDocument();
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('disables styles', () => {
		const rendered = render(<Element type={'div'} disableStyles />);

		const element = rendered.container.querySelector('.ss__element');
		expect(element?.classList).toHaveLength(1);
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('is themeable with ThemeProvider', () => {
		const globalTheme = {
			components: {
				element: {
					className: 'theme-class',
				},
			},
		};

		const rendered = render(
			<ThemeProvider theme={globalTheme}>
				<Element type="div" />
			</ThemeProvider>
		);

		const dropdownElement = rendered.container.querySelector('.ss__element');
		expect(dropdownElement).toHaveClass(globalTheme.components.element.className);
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('is themeable with theme prop', () => {
		const propTheme = {
			components: {
				element: {
					className: 'theme-class',
				},
			},
		};

		const rendered = render(<Element type="div" theme={propTheme} />);

		const dropdownElement = rendered.container.querySelector('.ss__element');
		expect(dropdownElement).toHaveClass(propTheme.components.element.className);
		expect(rendered.asFragment()).toMatchSnapshot();
	});

	it('is themeable and theme prop overrides ThemeProvider', () => {
		const globalTheme = {
			components: {
				element: {
					className: 'global-class',
				},
			},
		};

		const propTheme = {
			components: {
				element: {
					className: 'theme-class',
				},
			},
		};

		const rendered = render(
			<ThemeProvider theme={globalTheme}>
				<Element type="div" theme={propTheme} />
			</ThemeProvider>
		);

		const dropdownElement = rendered.container.querySelector('.ss__element');
		expect(dropdownElement).toHaveClass(propTheme.components.element.className);
		expect(dropdownElement).not.toHaveClass(globalTheme.components.element.className);
		expect(rendered.asFragment()).toMatchSnapshot();
	});
});
