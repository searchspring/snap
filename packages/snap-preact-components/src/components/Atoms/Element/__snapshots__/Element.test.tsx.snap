// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element Component disables styles 1`] = `
<DocumentFragment>
  <div
    class="ss__element"
  />
</DocumentFragment>
`;

exports[`Element Component doesnt render without type prop 1`] = `<DocumentFragment />`;

exports[`Element Component ignores onClick attribute 1`] = `
<DocumentFragment>
  <div
    attribute2="two"
    attribute3="3"
    class="ss__element emotion-0"
  >
    some text
  </div>
</DocumentFragment>
`;

exports[`Element Component is themeable and theme prop overrides ThemeProvider 1`] = `
<DocumentFragment>
  <div
    class="ss__element theme-class emotion-0"
  />
</DocumentFragment>
`;

exports[`Element Component is themeable with ThemeProvider 1`] = `
<DocumentFragment>
  <div
    class="ss__element theme-class emotion-0"
  />
</DocumentFragment>
`;

exports[`Element Component is themeable with theme prop 1`] = `
<DocumentFragment>
  <div
    class="ss__element theme-class emotion-0"
  />
</DocumentFragment>
`;

exports[`Element Component renders as div 1`] = `
<DocumentFragment>
  <div
    class="ss__element emotion-0"
  >
    some text
  </div>
</DocumentFragment>
`;

exports[`Element Component renders as label 1`] = `
<DocumentFragment>
  <label
    class="ss__element emotion-0"
  >
    some text
  </label>
</DocumentFragment>
`;

exports[`Element Component renders as p 1`] = `
<DocumentFragment>
  <p
    class="ss__element emotion-0"
  >
    some text
  </p>
</DocumentFragment>
`;

exports[`Element Component renders as span 1`] = `
<DocumentFragment>
  <span
    class="ss__element emotion-0"
  >
    some text
  </span>
</DocumentFragment>
`;

exports[`Element Component renders with attributes 1`] = `
<DocumentFragment>
  <div
    attribute1="one"
    attribute2="two"
    attribute3="3"
    class="ss__element emotion-0"
  >
    some text
  </div>
</DocumentFragment>
`;

exports[`Element Component renders with className prop 1`] = `
<DocumentFragment>
  <div
    class="ss__element classy emotion-0"
  />
</DocumentFragment>
`;

exports[`Element Component uses bocachica theme 1`] = `
<DocumentFragment>
  .emotion-0 {
  color: #6d7175;
}

<div
    class="ss__element emotion-0"
  >
    some text
  </div>
</DocumentFragment>
`;
